### ISSUE TITLE
Enable MSBuild Server for all dotnet CLI-based builds by default

### ISSUE THREAD

**Issue by JanKrivanek** [dotnet/msbuild#9379]
MSBuild Server keeps the main MSBuild process alive between builds, so the next build skips process startup and reuses its caches. The worker MSBuild processes already has such an optimization, but now the main process will also benefit from it. This is particularly beneficial for small solution warm build, as for them considerable part of the build was executing in the main process for which caches are lost. 

Multithreaded mode depends on it too: for MT mode all of the caches lives in the main process, unlike the non-MT mode, so keeping it alive will ensure MT would not detriment performance of warm builds. Server process is also the only place MSBuild can turn on Server GC, which multithreaded builds need. This epic turns the server on by default for dotnet CLI builds.

## Steps

- [x] [#11355](https://github.com/dotnet/msbuild/issues/11355) [#11356](https://github.com/dotnet/msbuild/issues/11356) [#11297](https://github.com/dotnet/msbuild/issues/11297) Enable MSBuild Server in the MSBuild repo CI and local development
- [x] [#11357](https://github.com/dotnet/msbuild/issues/11357) Onboard partner .NET repositories
- [x] [#13315](https://github.com/dotnet/msbuild/issues/13315) Run Restore in a task host to work around the NuGet auth problem
- [ ] [#12246](https://github.com/dotnet/msbuild/issues/12246) Document behavioral changes and how static state behaves when a build process is reused [0.5 dev weeks]
- [x] [#11358](https://github.com/dotnet/msbuild/issues/11358) Enable MSBuild Server by default in the .NET SDK

## Stakeholders

.NET SDK CLI users, the VMR and partner CI, and task authors whose tasks keep static state between builds.

## Definition of done

Dotnet CLI builds use MSBuild Server by default.

<details>
<summary>Original description</summary>

### Context

MSBuild server is currently opt-in. There might be scenarios where it might bring performance gains. There as well might be possibility to further leverage MSBuild Server (though there is a overlap with current VS build caching techniques). The value and trade-off might not be fully clear

### Goal

* Collect feedback from stakeholders (PM/EM), suggest the path forward (if any) and secure stakeholders buy-in for those
* If needed actions are identified - start properly tracking and prioritizing those

### Current State

We're concerned about runway for getting MSBuild Server enabled for all SDK users in the 10 GA release. We know of one area of concern ([NuGet auth](https://github.com/dotnet/msbuild/issues/12246)) that we need to do a mitigation around, and are concerned that we don't have enough previews/RCs left to 
* 
* roll out the change
* get broad user feedback/experience
* react as necessary to that feedback

without either running out of time or negatively impacting the user perception of 10 GA's stability.
 
Therefore we're going to take the following course of action:

* 10 GA:
  * mitigate the NuGet auth problems by pushing the Restore task out of proc when MSBuild Server is enabled (also helps folks that use server today)
  * get primary dotnet-org repos using Server to kick the tires
* 10.0.200
  * enable server at the SDK level

This mirrors our plan for the multithreaded work, where we're doing enablement/foundational work for 10 GA but not enabling it by default due to time/risk.  

</details>

**Comment by baronfel** [comment]
One of the reasons it's off by default is a [known issue](https://github.com/NuGet/Home/issues/12540) with the NuGet credential providers used in Restore - which has been fixed as of a couple weeks ago! We should be clear to pick this effort back up now.

**Comment by baronfel** [comment]
The general approach we should use here is:

* [ ] enabling Server for the MSBuild Repo and dogfooding for a bit
* [ ] enabling Server for dotnet/runtime, dotnet/aspnetcore, and dotnet/sdk as a second 'ring' of testing
* [ ] changing the default in the .NET SDK so that starting in a specific .NET SDK 10 preview release the default is back on. This will require a breaking change doc in the SDK change docs, but that is a well-understood process.

**Comment by SimaTian** [comment]
Status update for dogfooding:
 - MSBuild pipeline dogfooding - in progress, works without issues.
 - MSBuild internal CLI dogfooding 
    - in progress, so far no big issues
    - minor annoyance that I need to disable it if I want to use VS profiler.
    - TODO: collect more feedback from the team.
 - ASP.NET - the build with server passed succesfully, however the follow up tests crashed. We need to figure out a way to disable MSBuild server for the test stage. (similar clash was happening for MSBuild pipeline)

Not yet started: 
 - sdk
 - runtime

**Comment by SimaTian** [comment]
ASPnet manual run run into some issues - I'm unsure if my way of setting up the variable in the pipeline is reasonable as it would probably affect all the other builds. And when removing it early, it appears that the build run without the variable or at least it was hard to validate.

To avoid these pitfalls I've prepared a PR so that I have more control over what is happening.
https://github.com/dotnet/aspnetcore/pull/61993

**Comment by JanProvaznik** [comment]
# MSBuild Server default-on — final consolidated state (post #13660 + #13716 merge)

**Update 2026-05-14:** With #13660 (RestoreTask transient TaskHost) and #13716 (apphost `DOTNET_ROOT` + caller-level `TimeoutException` catch) merged, I re-audited my prototype branch's remaining "fix" candidates. Applying three filters that I should have applied originally:

1. **Worker-vs-main filter** — worker-node statics already accumulate today via `/nr:true` reuse, so they're not a server regression.
2. **Contract-violation filter** — if the only way to trigger the leak is for a task or logger to violate the MSBuild contract, the engine shouldn't ship the fix.
3. **Currently-observable filter** — "could matter in a future scenario" doesn't justify a PR.

**Result: every remaining prototype fix candidate fails one of the three filters.** No additional MSBuild engine changes are blocking server default-on beyond what already merged.

## Per-candidate disposition

| Candidate | Disposition | Reason |
|---|---|---|
| Bump 1 s hot-server connect timeout to 5 s | DROP | Speculative — never measured the recycle-gap duration. PR #13716's caller-level `TimeoutException` catch already prevents the crash. Bumping a hardcoded number without measurement is hygiene without evidence. |
| `BinaryLogger.Shutdown` unsubscribe `AnyEventRaised` | DROP | Doesn't fire under `dotnet build` server: `XMake.cs:3545` constructs a fresh `BinaryLogger` per request. Only triggers for programmatic API consumers reusing a logger across `BuildManager.BeginBuild`/`EndBuild` cycles — not the server scenario. |
| `BuildManager.ProjectCacheDescriptors` clear on `EndBuild` | DROP | Code path that adds entries (`Evaluator.CollectProjectCachePlugins`) is gated on `RunningInVisualStudio`. MSBuild server isn't VS. Latent risk if VS ever hosts MSBuild server. |
| `CommandLineSwitches.SwitchesFromResponseFiles` clear | DROP | False alarm in my W1b inventory — already cleared in `CommandLineParser.ResetGatheringSwitchesState():724-728` on every parse. |
| `ConsoleConfiguration.Provider` set per request, never restored | DROP | Same WITHDRAWN reasoning as PR #13757 — only manifests via contract-violating loggers. |
| `CurrentCulture`/`UICulture` reset between requests | DROP | Same WITHDRAWN reasoning as PR #13756 — explicit per-request set already covers the legitimate path. |
| TEL-1 `IsServerModeEnabled` telemetry field | DEFER | Not a fix; adoption-rate monitoring. File separately if/when the team wants telemetry on server adoption (independent of any blocker). |
| #13758 `-mt` implies MSBuild Server | KEEP (open) | Not a fix; design proposal for enablement. Already an open draft PR. |

## Out-of-tree task statics — unchanged status

- **`Microsoft.DotNet.SdkResolver.NETCoreSdkResolver.s_compatibleSdks`** (dotnet/sdk) — main-node-only static cache, real new-class server regression. Belongs to dotnet/sdk; would need an `IStatefulSdkResolver` callback wired to `BuildManager.ResetCaches()`.
- **`Microsoft.NET.Build.Containers.CreateNewImage`** env-var credential mutation (dotnet/sdk) — pre-existing in-process credential leak; bigger concern with `/mt` than server.
- **NuGet `GetRestoreSettingsTask._machineWideSettings`** — same pattern as `RestoreTask`; could be added to PR #13660's allow-list.

These were always tracked as out-of-tree concerns; nothing changed for them.

## Validation plan for default-on

1. ✅ **#13660 merged** — RestoreTask transient TaskHost
2. ✅ **#13716 merged** — apphost `DOTNET_ROOT` + caller-level `TimeoutException` catch
3. **Optionally merge #13758** — `-mt` implies server (enablement decision, not a bug fix)
4. **VMR-side per-repo isolation** — `MSBUILDNODEHANDSHAKESALT=$(RepositoryName)` per repo `<Exec>` in `repo-projects/Directory.Build.targets` (zero MSBuild code change)
5. **Out-of-tree work** — dotnet/sdk `NETCoreSdkResolver` cache invalidation (separately filed there)

## Lesson learned

My W1b/W1c/W1d/W1e inventories were structurally classified ("static field exists, event subscription not paired, env var mutated") but did not apply the three filters above. A proper audit starts with: (a) is this code path exercised by the actual `dotnet build` server scenario; (b) does this fail today *without server* under `/nr:true` worker reuse; (c) does the failure require a contract-respecting program. Future server-mode audits should start with those three filters before doing any code-shape inventory.

The full per-candidate analysis is in `session-notes/rec-9-mainnode-statics.md` and `session-notes/rec-final-empty-result.md` on the prototype branch.

---

*Comment last edited 2026-05-14 to reflect post-merge state. Prior versions claimed multiple "blocking fixes" that on closer audit either don't apply, are already done, or would defend against contract violations. Apologies for the noise.*
